src/i2s.rs

```diff
@@ -0,0 +1,1835 @@
+        Din: MaybePin + 'static,
+        Dout: MaybePin + 'static,
+        Mclk: MaybePin + 'static,
+        Ws: Pin + IOPin + 'static,
+    {
+        /// MCLK (master/controller clock) pin. Optional output.
+        mclk: Mclk,
+
+        /// BCLK (bit clock) pin. Input in target/slave mode, output in controller/master mode.
+        bclk: PeripheralRef<'static, Bclk>,
+
+        /// WS (word select) pin. Input in target/slave mode, output in controller/master mode.
+        ws: PeripheralRef<'static, Ws>,
+
+        /// Data output pin.
+        data_out: Dout,
+
+        /// Data input pin.
+        data_in: Din,
+
+        /// Invert the MCLK signal.
+        mclk_invert: bool,
+
+        /// Invert the BCLK signal.
+        bclk_invert: bool,
+
+        /// Invert the WS signal.
+        ws_invert: bool,
+    }
+
+    impl<Bclk, Din, Dout, Mclk, Ws> StdGpioConfig<Bclk, Din, Dout, Mclk, Ws>
+    where
+        Bclk: Pin + IOPin + 'static,
+        Din: MaybePin + 'static,
+        Dout: MaybePin + 'static,
+        Mclk: MaybePin + 'static,
+        Ws: Pin + IOPin + 'static,
+    {
+        /// Convert to the ESP-IDF SDK `i2s_std_gpio_config_t` representation.
+        pub(crate) fn as_sdk(&self) -> i2s_std_gpio_config_t {
+            let invert_flags = i2s_std_gpio_config_t__bindgen_ty_1 {
+                _bitfield_1: i2s_std_gpio_config_t__bindgen_ty_1::new_bitfield_1(
+                    self.mclk_invert as u32,
+                    self.bclk_invert as u32,
+                    self.ws_invert as u32,
+                ),
+                ..Default::default()
+            };
+
+            i2s_std_gpio_config_t {
+                mclk: self.mclk.pin(),
+                bclk: self.bclk.pin(),
+                ws: self.ws.pin(),
+                dout: self.data_out.pin(),
+                din: self.data_in.pin(),
+                invert_flags,
+            }
+        }
+    }
+
+    /// Incremental builder for a standard mode GPIO configuration.
+    pub struct StdGpioConfigBuilder<Bclk, Din, Dout, Mclk, Ws> {
+        bclk: Bclk,
+        data_in: Din,
+        data_out: Dout,
+        mclk: Mclk,
+        ws: Ws,
+        bclk_invert: bool,
+        mclk_invert: bool,
+        ws_invert: bool,
+    }
+
+    impl Default for StdGpioConfigBuilder<(), (), (), (), ()> {
+        fn default() -> Self {
+            Self {
+                bclk: (),
+                data_in: (),
+                data_out: (),
+                mclk: (),
+                ws: (),
+                bclk_invert: false,
+                mclk_invert: false,
+                ws_invert: false,
+            }
+        }
+    }
+
+    impl<Din, Dout, Mclk, Ws> StdGpioConfigBuilder<(), Din, Dout, Mclk, Ws> {
+        /// Set the bit clock (BCK) pin.
+        ///
+        /// This must be called before calling [StdGpioConfigBuilder::build].
+        #[must_use]
+        #[inline(always)]
+        pub fn bclk<Bclk: Pin + IOPin + 'static>(
```

Comment:
I strongly suggest we retire these structs and builders actually, but just to point why you can't reuse pins: all other drivers take e.g. `bclk: impl Peripheral<P = Bclk> + 'd` which allows you to pass a pin peripheral either by moving it (and then `'d` essentially becomes `'static` or by `&mut` ref-ing it, and then `'d` models the lifetime of the mut ref.